src/shardeum/secureAccounts.ts

```diff
@@ -0,0 +1,327 @@
+    return { success: false, reason: 'Transfer not allowed yet, time restriction' }
+  }
+
+  if (transferAmount > secureAccount.data.nextTransferAmount) {
+    return { success: false, reason: 'Transfer amount exceeds allowed limit' }
+  }
+
+  return { success: true, reason: 'Valid transaction' }
+}
+
+export async function apply(
+  tx: InternalTx,
+  txId: string,
+  txTimestamp: number,
+  wrappedStates: WrappedStates,
+  shardus: Shardus,
+  applyResponse: ShardusTypes.ApplyResponse
+): Promise<void> {
+  const secureAccountConfig = secureAccountDataMap.get(tx.accountName)
+
+  const sourceEOA = wrappedStates[toShardusAddress(secureAccountConfig.SourceFundsAddress, AccountType.Account)];
+  const destEOA = wrappedStates[toShardusAddress(secureAccountConfig.RecipientFundsAddress, AccountType.Account)];
+  const secureAccount = wrappedStates[toShardusAddress(secureAccountConfig.SecureAccountAddress, AccountType.SecureAccount)];
+
+  // throw if any of the required accounts are not found
+  if (!sourceEOA || !destEOA || !secureAccount) {
+    console.log('One or more required accounts not found', sourceEOA, destEOA, secureAccount)
+    throw new Error('One or more required accounts not found');
+  }
+
+  const sourceEOAData = sourceEOA.data as WrappedEVMAccount;
+  const destEOAData = destEOA.data as WrappedEVMAccount;
+  const secureAccountData = secureAccount.data as SecureAccount;
+
+  const amount = BigInt(tx.amount);
+
+  sourceEOAData.account.balance = BigInt(sourceEOAData.account.balance) - amount
+  sourceEOAData.account.nonce += BigInt(1)
+  destEOAData.account.balance = BigInt(destEOAData.account.balance) + amount
+
+  // update timestamp for each account
+  sourceEOAData.timestamp = txTimestamp;
+  destEOAData.timestamp = txTimestamp;
+  secureAccountData.timestamp = txTimestamp;
+
+  secureAccountData.nonce += 1;
+
+  updateEthAccountHash(sourceEOAData);
+  updateEthAccountHash(destEOAData);
+  updateEthAccountHash(secureAccountData);
+  
+  const wrappedSourceEOA = _shardusWrappedAccount(sourceEOAData);
+  const wrappedDestEOA = _shardusWrappedAccount(destEOAData);
+  const wrappedSecureAccount = _shardusWrappedAccount(secureAccountData);
+
+  shardus.applyResponseAddChangedAccount(
+    applyResponse,
+    toShardusAddress(secureAccountConfig.SourceFundsAddress, AccountType.Account),
+    wrappedSourceEOA as ShardusTypes.WrappedResponse,
+    txId,
+    applyResponse.txTimestamp
+  );
+  shardus.applyResponseAddChangedAccount(
+    applyResponse,
+    toShardusAddress(secureAccountConfig.RecipientFundsAddress, AccountType.Account),
+    wrappedDestEOA as ShardusTypes.WrappedResponse,
+    txId,
+    applyResponse.txTimestamp
+  );
+  shardus.applyResponseAddChangedAccount(
+    applyResponse,
+    toShardusAddress(secureAccountConfig.SecureAccountAddress, AccountType.SecureAccount),
+    wrappedSecureAccount as ShardusTypes.WrappedResponse,
+    txId,
+    applyResponse.txTimestamp
+  );
```

Comment:
applyResponseAddChangedAccount if will fail for any account, should be rolled back to initial state.

Reply:
hmm yeah thats a good call thanks

Reply:
I will make a followup branch to add the checkpointing / atomicity since it will require some more testing. Thanks for pointing that out!